homeassistant/components/onvif/__init__.py

```diff
@@ -155,3 +158,32 @@ async def async_populate_options(hass: HomeAssistant, entry: ConfigEntry) -> Non
     }
 
     hass.config_entries.async_update_entry(entry, options=options)
+
+
+@callback
+def _migrate_camera_entities_unique_ids(
+    hass: HomeAssistant, config_entry: ConfigEntry, device: ONVIFDevice
+) -> None:
+    """Migrate unique ids of camera entities from profile index to profile token."""
+    entity_reg = er.async_get(hass)
+    entities: list[er.RegistryEntry] = er.async_entries_for_config_entry(
+        entity_reg, config_entry.entry_id
+    )
+    mac_or_serial = device.info.mac or device.info.serial_number
+    for entity in entities:
+        if entity.domain == "camera":
+            old_uid_start = f"{mac_or_serial}_"
+            new_uid_start = f"{mac_or_serial}#"
+            if (
+                entity.unique_id.startswith(old_uid_start)
+                or entity.unique_id == mac_or_serial
+            ):
+                index = 0
+                if entity.unique_id.startswith(old_uid_start):
+                    index = int(entity.unique_id[len(old_uid_start) :])
```

Comment:
please add a try/except around the call to `int` in case it fails to parse for any reason